### PR TITLE
Exclude filter columns from projection

### PR DESCRIPTION
**Pull request by eddyxu**
In the case for `dataset.scanner(columns=["foo"], filters=(col("bar") == 1)`, the result should only have column `foo`.

### CODE REVIEW

File: cpp/src/lance/io/reader.cc [hunk 1]
```diff
@@ -289,6 +289,7 @@ ::arrow::Result<std::shared_ptr<::arrow::RecordBatch>> FileReader::ReadBatch(
 
 ::arrow::Result<std::shared_ptr<::arrow::RecordBatch>> FileReader::ReadBatch(
     const lance::format::Schema& schema, int32_t batch_id, const ArrayReadParams& params) const {
+  assert(!schema.fields().empty());
```
Comment (changhiskhan) [inline cpp/src/lance/io/reader.cc:292]: Does this raise exception if false or just segfault?

File: cpp/src/lance/io/exec/take.cc [hunk 2]
```diff
@@ -14,6 +14,10 @@
 
 #include "lance/io/exec/take.h"
 
+#include <arrow/result.h>
```
Comment (changhiskhan) [inline cpp/src/lance/io/exec/take.cc:17]: Where r they used?

File: cpp/src/lance/io/exec/project.cc [hunk 3]
```diff
@@ -70,14 +72,21 @@ ::arrow::Result<std::unique_ptr<Project>> Project::Make(
...
+
   assert(child_);
-  return child_->Next();
+  ARROW_ASSIGN_OR_RAISE(auto batch, child_->Next());
+  if (batch.eof()) {
+    return batch;
+  }
+  ARROW_ASSIGN_OR_RAISE(auto projected_batch,
+                        lance::arrow::ApplyProjection(batch.batch, *projected_schema_));
+  return ScanBatch(projected_batch, batch.batch_id, batch.indices);
```
Comment (changhiskhan) [inline cpp/src/lance/io/exec/project.cc:89]: Maybe makes sense to add ScanBatch.Project ?

**Review by changhiskhan — APPROVED** [review]
Some nits

File: cpp/src/lance/arrow/scanner.cc [hunk 4]
```diff
@@ -67,37 +65,18 @@ ::arrow::Result<std::shared_ptr<::arrow::dataset::Scanner>> ScannerBuilder::Fini
...
-        if (std::find(std::begin(columns), std::end(columns), filtered_col_name) == columns.end()) {
-          columns.emplace_back(filtered_col_name);
-        }
-      }
-    }
     ARROW_ASSIGN_OR_RAISE(auto projected_schema, schema.Project(columns));
-    scanner->options()->dataset_schema = projected_schema->ToArrow();
     ARROW_ASSIGN_OR_RAISE(scanner->options()->filter,
                           scanner->options()->filter.Bind(*scanner->options()->dataset_schema));
+    scanner->options()->projected_schema = projected_schema->ToArrow();
```
Comment (changhiskhan) [inline cpp/src/lance/arrow/scanner.cc:71]: Does it matter whether this is done before or after the filter bind?

File: cpp/src/lance/io/exec/take.cc (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/io/exec/take.cc:17]: In this same file, it uses arrow result multiple times.

Including the headers in `.cc` files can allow `.h` to just do forward declaration in general.

File: cpp/src/lance/io/reader.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/io/reader.cc:292]: good point. it does segfault in debug mode. 

I considered that this should be a condition that never happens, thus as a bug.

File: cpp/src/lance/arrow/scanner.cc (hunk 4 above)
Comment (eddyxu) [reply cpp/src/lance/arrow/scanner.cc:71]: the filter bined on dataset schema, which changes only the filter internal.

This assigntment is over project schema, which is separated from dataset schema. I think it does not matter to the filter binding order.

File: cpp/src/lance/io/exec/project.cc (hunk 3 above)
Comment (eddyxu) [reply cpp/src/lance/io/exec/project.cc:89]: Sure, I can move this to ScanBatch.

**Pull request merged**